packages/workbox-background-sync/src/lib/QueueDb.ts

```diff
@@ -0,0 +1,149 @@
+/*
+  Copyright 2018 Google LLC
+
+  Use of this source code is governed by an MIT-style
+  license that can be found in the LICENSE file or at
+  https://opensource.org/licenses/MIT.
+*/
+
+import {openDB, DBSchema, IDBPDatabase} from 'idb';
+import {RequestData} from './StorableRequest.js';
+
+interface QueueDBSchema extends DBSchema {
+  requests: {
+    key: number;
+    value: UnidentifiedQueueStoreEntry;
+    indexes: {queueName: string};
+  };
+}
+
+const DB_VERSION = 3;
+const DB_NAME = 'workbox-background-sync';
+const OBJECT_STORE_NAME = 'requests';
+const INDEXED_PROP = 'queueName';
+
+export interface UnidentifiedQueueStoreEntry {
+  requestData: RequestData;
+  timestamp: number;
+  id?: number;
+  queueName?: string;
+  metadata?: object;
+}
+
+export interface QueueStoreEntry extends UnidentifiedQueueStoreEntry {
+  id: number;
+}
+
+/**
+ * A class to interact directly with IndexedDB to store a retrieve queue requests.
+ *
+ * @private
+ */
+
+export class QueueDb {
+  private _db: IDBPDatabase<QueueDBSchema> | null = null;
+
+  /**
+   * Opens a connection to IDBDatabase.
+   *
+   * @private
+   */
+  private async open() {
+    if (!this._db) {
+      this._db = await openDB(DB_NAME, DB_VERSION, {
+        upgrade: this._upgradeDb,
+      });
+    }
+  }
+
+  /**
+   *
+   * @param {UnidentifiedQueueStoreEntry} entry
+   */
+  async addEntry(entry: UnidentifiedQueueStoreEntry) {
+    await this.open();
+    await this._db!.add(OBJECT_STORE_NAME, entry);
+  }
+
+  /**
+   * Returns the first entry in the ObjectStore
+   *
+   * @return {QueueStoreEntry}
+   */
+  async getFirstEntry(): Promise<QueueStoreEntry | any> {
+    await this.open();
+    const cursor = await this._db
+      ?.transaction(OBJECT_STORE_NAME)
+      .store.openCursor();
+
+    if (cursor) {
+      return cursor.value as QueueStoreEntry;
```

Comment:
I'm kind of surprised that you have to cast `cursor.value` to `QueueStoreEntry` here—what is the type of `cursor.value` if you don't cast? One of the advantages of the IDB library, I thought, was that it "knew" about the types defined in `QueueDBSchema`.

Reply:
There are two interfaces defined for an entry UnidentifiedQueueStoreEntry and QueueStoreEntry, the difference is that QueueStoreEntry has to have an id, for the UnidentifiedQueueStoreEntry the id is optional. 

To add a record to it has to be unidentified because IDB sets the for the record when it. So the type is specified in the schema as UnidentifiedQueueEntry, line 16. The methods return Promise<UnidentifiedQueryEntry> that is why the need for the casting. No sure if there is a better way to handle that.